#!/home/gloria/anaconda3/envs/py3_8_env/bin/python
import os
import time
import numpy as np
import cv2
import torch
import sys

sys.path.append('/home/gloria/catkin_ws/src/yolov5')  # absolute path to YOLOV5
sys.path.append('/home/zjrobot/SLAM/detect/yolov5')  # absolute path to YOLOV5
from models.experimental import attempt_load


def box_iou(box1, box2):
    # https://github.com/pytorch/vision/blob/master/torchvision/ops/boxes.py
    """
    Return intersection-over-union (Jaccard index) of boxes.
    Both sets of boxes are expected to be in (x1, y1, x2, y2) format.
    Arguments:
        box1 (Tensor[N, 4])
        box2 (Tensor[M, 4])
    Returns:
        iou (Tensor[N, M]): the NxM matrix containing the pairwise
            IoU values for every element in boxes1 and boxes2
    """

    def box_area(box):
        # box = 4xn
        return (box[2] - box[0]) * (box[3] - box[1])

    area1 = box_area(box1.T)
    area2 = box_area(box2.T)

    # inter(N,M) = (rb(N,M,2) - lt(N,M,2)).clamp(0).prod(2)
    inter = (torch.min(box1[:, None, 2:], box2[:, 2:]) - torch.max(box1[:, None, :2], box2[:, :2])).clamp(0).prod(2)
    return inter / (area1[:, None] + area2 - inter)  # iou = inter / (area1 + area2 - inter)

def xywh2xyxy(x):
    # Convert nx4 boxes from [x, y, w, h] to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    y = torch.zeros_like(x) if isinstance(x, torch.Tensor) else np.zeros_like(x)
    y[:, 0] = x[:, 0] - x[:, 2] / 2  # top left x
    y[:, 1] = x[:, 1] - x[:, 3] / 2  # top left y
    y[:, 2] = x[:, 0] + x[:, 2] / 2  # bottom right x
    y[:, 3] = x[:, 1] + x[:, 3] / 2  # bottom right y
    return y

def non_max_suppression(prediction, conf_thres=0.1, iou_thres=0.6, merge=False, classes=None, agnostic=False):
    """Performs Non-Maximum Suppression (NMS) on inference results

    Returns:
         detections with shape: nx6 (x1, y1, x2, y2, conf, cls)
    """

    nc = prediction[0].shape[1] - 5  # number of classes
    xc = prediction[..., 4] > conf_thres  # candidates

    # Settings
    min_wh, max_wh = 2, 4096  # (pixels) minimum and maximum box width and height
    max_det = 300  # maximum number of detections per image
    time_limit = 10.0  # seconds to quit after
    redundant = True  # require redundant detections
    multi_label = nc > 1  # multiple labels per box (adds 0.5ms/img)

    t = time.time()
    output = [None] * prediction.shape[0]
    for xi, x in enumerate(prediction):  # image index, image inference
        # Apply constraints
        # x[((x[..., 2:4] < min_wh) | (x[..., 2:4] > max_wh)).any(1), 4] = 0  # width-height
        x = x[xc[xi]]  # confidence

        # If none remain process next image
        if not x.shape[0]:
            continue

        # Compute conf
        x[:, 5:] *= x[:, 4:5]  # conf = obj_conf * cls_conf

        # Box (center x, center y, width, height) to (x1, y1, x2, y2)
        box = xywh2xyxy(x[:, :4])

        # Detections matrix nx6 (xyxy, conf, cls)
        if multi_label:
            i, j = (x[:, 5:] > conf_thres).nonzero(as_tuple=False).T
            x = torch.cat((box[i], x[i, j + 5, None], j[:, None].float()), 1)
        else:  # best class only
            conf, j = x[:, 5:].max(1, keepdim=True)
            x = torch.cat((box, conf, j.float()), 1)[conf.view(-1) > conf_thres]

        # Filter by class
        if classes:
            x = x[(x[:, 5:6] == torch.tensor(classes, device=x.device)).any(1)]

        # Apply finite constraint
        # if not torch.isfinite(x).all():
        #     x = x[torch.isfinite(x).all(1)]

        # If none remain process next image
        n = x.shape[0]  # number of boxes
        if not n:
            continue

        # Sort by confidence
        # x = x[x[:, 4].argsort(descending=True)]

        # Batched NMS
        c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
        boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
        i = torch.ops.torchvision.nms(boxes, scores, iou_thres)
        if i.shape[0] > max_det:  # limit detections
            i = i[:max_det]
        if merge and (1 < n < 3E3):  # Merge NMS (boxes merged using weighted mean)
            try:  # update boxes as boxes(i,4) = weights(i,n) * boxes(n,4)
                iou = box_iou(boxes[i], boxes) > iou_thres  # iou matrix
                weights = iou * scores[None]  # box weights
                x[i, :4] = torch.mm(weights, x[:, :4]).float() / weights.sum(1, keepdim=True)  # merged boxes
                if redundant:
                    i = i[iou.sum(1) > 1]  # require redundancy
            except:  # possible CUDA error https://github.com/ultralytics/yolov3/issues/1139
                print(x, i, x.shape, i.shape)
                pass

        output[xi] = x[i]
        if (time.time() - t) > time_limit:
            break  # time limit exceeded

    return output

def letterbox(img, new_shape=(640, 640), color=(114, 114, 114), auto=True, scaleFill=False, scaleup=True):
    # Resize image to a 32-pixel-multiple rectangle https://github.com/ultralytics/yolov3/issues/232
    shape = img.shape[:2]  # current shape [height, width]
    if isinstance(new_shape, int):
        new_shape = (new_shape, new_shape)

    # Scale ratio (new / old)
    r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])
    if not scaleup:  # only scale down, do not scale up (for better test mAP)
        r = min(r, 1.0)

    # Compute padding
    ratio = r, r  # width, height ratios
    new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))
    dw, dh = new_shape[1] - new_unpad[0], new_shape[0] - new_unpad[1]  # wh padding
    if auto:  # minimum rectangle
        dw, dh = np.mod(dw, 64), np.mod(dh, 64)  # wh padding
    elif scaleFill:  # stretch
        dw, dh = 0.0, 0.0
        new_unpad = (new_shape[1], new_shape[0])
        ratio = new_shape[1] / shape[1], new_shape[0] / shape[0]  # width, height ratios

    dw /= 2  # divide padding into 2 sides
    dh /= 2

    if shape[::-1] != new_unpad:  # resize
        img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_LINEAR)
    top, bottom = int(round(dh - 0.1)), int(round(dh + 0.1))
    left, right = int(round(dw - 0.1)), int(round(dw + 0.1))
    img = cv2.copyMakeBorder(img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=color)  # add border
    return img, ratio, (dw, dh)

class ToyCar():
    def __init__(self,model_path =None ,imgsz =640, far_conf_thres = 0.3,near_conf_thres = 0.2,iou_thres = 0.3,model_half=False,device='gpu'):
        if model_path ==None:
            model_path =os.path.dirname(__file__) + 'model/best.pt'
        elif model_path.startswith('/'):
            assert os.path.exists(model_path)
        else:
            model_path = os.path.join(os.path.dirname(__file__),model_path)
        self.imgsz = imgsz
        self.far_conf_thres = far_conf_thres
        self.near_conf_thres = near_conf_thres
        self.iou_thres = iou_thres
        if device=='gpu':
            self.device = torch.device('cuda:0')
        else:
            self.device = torch.device('cpu')
        self.half = model_half
        self.model = attempt_load(model_path, map_location=self.device)#torch.load(model_path, map_location=self.device)['model'].float().fuse().eval() # to FP32
        if self.half:
            self.model.half()  # to FP16

    def run(self,img,is_near = False):
        if img is None:
            return [],[]
        img_size = img.shape[:2]
        img = self.precessing(img)
        img_rs = img.shape[2:]
        ratio = np.array([img_size[1]/img_rs[1],img_size[0]/img_rs[0],img_size[1]/img_rs[1],img_size[0]/img_rs[0]])
        with torch.no_grad():
            box_conf = self.model(img)[0]
        box,conf = self.postcessing(box_conf,is_near)
        if len(box)>0:
            box = (np.array(box)*ratio[None]).tolist()
        return box,conf

    def postcessing(self,pred, is_near=False):
        if is_near:
            conf_thres = self.near_conf_thres
        else:
            conf_thres = self.far_conf_thres
        pred = non_max_suppression(pred, conf_thres, self.iou_thres, classes=None, agnostic=False)[0]
        if pred is None:return [],[]
        box_conf = pred.cpu().numpy()
        return box_conf[:,:4].tolist(), box_conf[:,4].tolist()

    def precessing(self,img):
        imgo = letterbox(img, new_shape=self.imgsz)[0]
        # Convert
        img = imgo[:,:,::-1].transpose(2, 0, 1)  # BGR to RGB,
        img = np.ascontiguousarray(img)

        img = torch.from_numpy(img).to(self.device)
        img = img.half() if self.half else img.float()  # uint8 to fp16/32
        img /= 255.0  # 0 - 255 to 0.0 - 1.0
        if img.ndimension() == 3:
            img = img.unsqueeze(0)
        return img

def test_video():
    cam_id = 4
    if len(sys.argv)>1:
        cam_id = int(sys.argv[1])
    print('use video ',cam_id)
    cap = cv2.VideoCapture(cam_id)
    detect = ToyCar('../model/best.pt')
    while 1:
        ret, img = cap.read()
        if not ret:break
        torch.cuda.synchronize()
        st = time.time()
        img = cv2.resize(img,(640,320))

        box,conf = detect.run(img)
        torch.cuda.synchronize()
        for b in box:
            cv2.rectangle(img, (int(b[0]), int(b[1])), (int(b[2]), int(b[3])), (0, 254, 0), 1)
        print((time.time() - st))
        cv2.imshow('debug', img)
        cv2.waitKey(1)

if __name__ == '__main__':
    test_video()

